src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs

```diff
@@ -344,141 +354,154 @@ private static unsafe bool AllowHyphenDuringParsing(NumberFormatInfo info)
+            private static bool IsDecimalChar(char ch)
             {
-                char *ret = MatchChars(p, pEnd, negativeSign);
-                if (ret == null && allowHyphenDuringParsing && p < pEnd && *p == '-')
-                {
-                    ret = p + 1;
-                }
+                return (ch >= '0') && (ch <= '9');
+            }
 
-                return ret;
+            private static bool IsHexadecimalChar(char ch)
+            {
+                return IsDecimalChar(ch)
+                    || ((ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F'));
             }
 
-            private static unsafe bool ParseNumber(ref char* str, char* strEnd, NumberStyles options, ref NumberBuffer number, StringBuilder? sb, NumberFormatInfo numfmt, bool parseDecimal)
+            private static unsafe bool IsSeparatorProcessing(NumberStates state,
+                char* p,
+                char* pEnd,
+                FormattingSpecification formattingSpec,
+                string mainSeparator,
+                string separatorForCurrency,
+                out char* next)
             {
-                Debug.Assert(str != null);
-                Debug.Assert(strEnd != null);
-                Debug.Assert(str <= strEnd);
+                bool returnValue;
 
-                const int StateSign = 0x0001;
-                const int StateParens = 0x0002;
-                const int StateDigits = 0x0004;
-                const int StateNonZero = 0x0008;
-                const int StateDecimal = 0x0010;
-                const int StateCurrency = 0x0020;
+                if (state.HasFlag(NumberStates.Decimal))
+                {
+                    next = null;
+                    returnValue = false;
+                }
+                else
+                {
+                    returnValue = (
+                        ((next = MatchChars(p, pEnd, mainSeparator)) != null)
+                        || (
+                            (formattingSpec.IsCurrencyParsing)
+                            &&
+                            (!state.HasFlag(NumberStates.Currency))
+                            &&
+                            ((next = MatchChars(p, pEnd, separatorForCurrency)) != null)
+                        )
+                    );
+                }
 
-                number.scale = 0;
-                number.sign = false;
-                string decSep;                  // Decimal separator from NumberFormatInfo.
-                string groupSep;                // Group separator from NumberFormatInfo.
-                string? currSymbol = null;      // Currency symbol from NumberFormatInfo.
+                return returnValue;
+            }
 
-                bool allowHyphenDuringParsing = AllowHyphenDuringParsing(numfmt);
+            private static unsafe bool IsSignProcessing(NumberStates state,
```

Comment:
I think `IsSignProcessing` and `IsSeperatorProcessing` should be renamed to say what they are doing